lib/pbench/test/unit/agent/test_utils.py

```diff
@@ -253,3 +260,286 @@ def test_methods(monkeypatch):
         assert lrh.is_local("2600::1"), "'2600::1' should be local"
         assert lrh.is_local("2600::0:0:1"), "'2600::0:0:1' should be local"
         assert not lrh.is_local("2600::3"), "'2600::3' should be remote"
+
+
+class Test_failed_tool_init:
+
+    mock_tms_failed_data = {
+        "testhost.example.com": {
+            "hostdata": {
+                "alias": "bond",
+                "all-fqdns": "agent.example.com bond.example.com",
+                "all-ip-addresses": "192.168.0.1 172.0.0.1 127.0.0.1",
+                "domain": "example.com",
+                "fqdn": "agent.example.com",
+                "ip-address": "172.0.0.1 127.0.0.1",
+                "nis": "hostname[mock]: Local domain name not set",
+                "short": "agent",
+            },
+            "hostname": "testhost.example.com",
+            "installs": {
+                "dcgm": [0, "dcgm tool (dcgm-exporter) properly installed"],
+                "mpstat": [1, "mpstat: pbench-sysstat-12.0.3-1 is not installed"],
+            },
+            "kind": "tm",
+            "label": "",
+            "pid": "NNNNN",
+            "seqno": "",
+            "sha1": "(unknown)",
+            "version": "(unknown)",
+            "tools": {
+                "dcgm": "--inst=/var/tmp/pbench-test-utils/opt/pbench-agent/unittest-scripts",
+                "mpstat": "--interval=42 --options=forty-two",
+            },
+            "noop_tools": [],
+            "persistent_tools": ["dcgm"],
+            "transient_tools": [],
+            "failed_tools": ["mpstat"],
+            "posted": None,
+        }
+    }
```

Comment:
Sure, I wasn't sure whether I should trim these down or not, So, I added all data.

Reply:
The way to think about unit testing is that we want the unit tests to exercise the code-under-test, driving it down every possible path (or, at least, specific paths) and ensuring that it always produces the appropriate behavior.  The mechanism required to do that doesn't always have match realistic operation.  For instance, we generally mock out all subroutine calls, both to enable us to execute the CUT in isolation as well as to allow us to control its operation and to check its behavior.  Similarly, in terms of the data supplied to the CUT, we don't need to provide realistic inputs (that's something that _functional_ testing will do); instead we only need to provide the specific inputs required to drive the intended execution of the CUT.

One problem with unit testing is that it can be difficult to understand (and even more difficult to "proofread").  So, keeping definitions small, defining nothing more than we need to, and keeping the definitions in logical, easy-to-find places (like at the top of each test scenario) helps a lot.

Reply:
The pared down data makes it much easier to see how the three test cases are different from each other and what they are testing -- thanks!